WorkflowSwiftUI/Sources/Store.swift

```diff
@@ -0,0 +1,453 @@
+
+        for (keyPath, childStore) in childStores {
+            if childStore.isInvalid(newModel) {
+                childStore.invalidate()
+                childStores[keyPath] = nil
+            } else {
+                childStore.setModel(newModel)
+            }
+        }
+
+        childModelAccesses = childModelAccesses.filter { _, access in
+            !access.isInvalid(newModel)
+        }
+    }
+
+    func invalidate() {
+        invalidated = true
+        for childStore in childStores.values {
+            childStore.invalidate()
+        }
+    }
+}
+
+// MARK: - Subscripting
+
+public extension Store {
+    subscript<T>(dynamicMember keyPath: KeyPath<State, T>) -> T {
+        state[keyPath: keyPath]
+    }
+
+    subscript<T>(dynamicMember keyPath: WritableKeyPath<State, T>) -> T {
+        get {
+            state[keyPath: keyPath]
+        }
+        set {
+            send(keyPath: keyPath, value: newValue)
+        }
+    }
+
+    subscript<Action>(dynamicMember keyPath: KeyPath<Model, Sink<Action>>) -> Sink<Action> {
+        model[keyPath: keyPath]
+    }
+
+    subscript<Value>(
+        state state: KeyPath<State, Value>,
+        send send: KeyPath<Model, (Value) -> Void>
+    ) -> Value {
+        get {
+            self.state[keyPath: state]
+        }
+        set {
+            model[keyPath: send](newValue)
+        }
+    }
+
+    subscript<Value, Action>(
+        state state: KeyPath<State, Value>,
+        sink sink: KeyPath<Model, Sink<Action>>,
+        action action: CaseKeyPath<Action, Value>
+    ) -> Value {
+        get {
+            self.state[keyPath: state]
+        }
+        set {
+            model[keyPath: sink].send(action(newValue))
+        }
+    }
+}
+
+// MARK: - Scoping
+
+public extension Store {
+    /// Holds a cached child store for a nested ObservableModel on this store's model.
+    internal struct ChildStore {
+        var store: Any
+        var setModel: (Model) -> Void
+        var isInvalid: (Model) -> Bool
+
+        private var _invalidate: () -> Void
+
+        init(
+            store: Store<some ObservableModel>,
+            setModel: @escaping (Model) -> Void,
+            isInvalid: @escaping (Model) -> Bool
+        ) {
+            self.store = store
+            self.setModel = setModel
+            self.isInvalid = isInvalid
+
+            self._invalidate = {
+                store.invalidate()
+            }
+        }
+
+        func invalidate() {
+            _invalidate()
+        }
+    }
+
+    /// Represents an access to the "wrapper" of a nested child store, such as an Optional or
+    /// collection type.
+    ///
+    /// Each nested model's scope will track its own mutations, but we use this to track mutations
+    /// to the wrapper itself, such as changes to a collection size.
+    internal struct ChildModelAccess {
+        var willSet: (Store<Model>) -> Void
+        var didSet: (Store<Model>) -> Void
+        var changed: (Model, Model) -> Bool
```

Comment:
Nit: `onChange`?

Reply:
`isChanged` might be more apt — this is called when the model updates to determine if we should register a mutation.